File: 265.PaintHouseII/BottomUpDP.java
class BottomUpDP {
  public int minCostII(int[][] costs) {

    if (costs.length == 0)
      return 0;
    int k = costs[0].length;
    int n = costs.length;

    for (int house = 1; house < n; house++) {
      for (int color = 0; color < k; color++) {
        int min = Integer.MAX_VALUE;
        for (int previousColor = 0; previousColor < k; previousColor++) {
          if (color == previousColor)
            continue;
          min = Math.min(min, costs[house - 1][previousColor]);
        }
        costs[house][color] += min;
      }
    }

    // Find the minimum in the last row.
    int min = Integer.MAX_VALUE;
    for (int c : costs[n - 1]) {
      min = Math.min(min, c);
    }
    return min;
  }
}

File: DynamicProgramming/256.PaintHouse/bottom_up_dp.js
/**
 * @param {number[][]} costs
 * @return {number}
 */
var minCost = function (cost) {
  const n = cost.length;

  for (let i = 1; i < n; i++) {
    cost[i][0] += Math.min(cost[i - 1][1], cost[i - 1][2]);
    cost[i][1] += Math.min(cost[i - 1][0], cost[i - 1][2]);
    cost[i][2] += Math.min(cost[i - 1][0], cost[i - 1][1]);
  }

  return Math.min(cost[n - 1][0], cost[n - 1][1], cost[n - 1][2]);
};
